subprojects/logging/src/main/java/org/gradle/internal/logging/console/LogGroupingOutputEventListener.java

```diff
@@ -0,0 +1,229 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+public class LogGroupingOutputEventListener extends BatchOutputEventListener {
+    static final int SCHEDULER_CHECK_PERIOD_MS = 5000;
+    public static final String EOL = SystemProperties.getInstance().getLineSeparator();
+    private final BatchOutputEventListener listener;
+    private final ScheduledExecutorService executor;
+    private final TimeProvider timeProvider;
+    private final Object lock = new Object();
+
+    // Allows us to map progress and complete events to start events
+    private final Map<OperationIdentifier, Object> progressIdToBuildOperationIdMap = new LinkedHashMap<OperationIdentifier, Object>();
+
+    // Maintain a hierarchy of all build operation ids, so that child operations are grouped under their ancestor
+    private final Map<Object, Object> buildOperationIdForest = new LinkedHashMap<Object, Object>();
+
+    // Event groups that are in-progress and have not been completed
+    private final Map<Object, ArrayList<CategorisedOutputEvent>> outputEventGroups = new LinkedHashMap<Object, ArrayList<CategorisedOutputEvent>>();
+    private Object lastRenderedOperationId;
+    private boolean needsHeader;
+
+    public LogGroupingOutputEventListener(BatchOutputEventListener listener, TimeProvider timeProvider) {
+        this(listener, Executors.newSingleThreadScheduledExecutor(), timeProvider);
+    }
+
+    LogGroupingOutputEventListener(BatchOutputEventListener listener, ScheduledExecutorService executor, TimeProvider timeProvider) {
+        this.listener = listener;
+        this.executor = executor;
+        this.timeProvider = timeProvider;
+    }
+
+    @Override
+    public void onOutput(OutputEvent event) {
+        synchronized (lock) {
+            if (event instanceof EndOutputEvent) {
+                renderAllGroups(outputEventGroups);
+                listener.onOutput(event);
+//                executor.shutdown();
+            } else if (event instanceof ProgressStartEvent) {
+                onStart((ProgressStartEvent) event);
+            } else if (event instanceof ProgressEvent) {
+                ProgressEvent progressEvent = (ProgressEvent) event;
+                groupOrForward(progressIdToBuildOperationIdMap.get(progressEvent.getProgressOperationId()), progressEvent);
+            } else if (event instanceof RenderableOutputEvent) {
+                RenderableOutputEvent renderableOutputEvent = (RenderableOutputEvent) event;
+                groupOrForward(renderableOutputEvent.getBuildOperationId(), renderableOutputEvent);
+            } else if (event instanceof ProgressCompleteEvent) {
+                onComplete((ProgressCompleteEvent) event);
+            } else {
+                listener.onOutput(event);
+            }
+
+            // TODO(ew): Use strategy from ThrottlingOEL to schedule runnable once every 5 seconds on new output
```

Comment:
I want to get this reviewed before adding this piece. I intend to start work on this today.